### DISCUSSION TITLE
Anharmonic potential

### DISCUSSION THREAD

**Discussion by LiuHaoXuan0403** [ttadano/alamode#214] [General]
Hi Dr. Tadano,

I hope this message finds you well.

I am currently working with ALAMODE to study anharmonic effects in materials. I would like to use second-order, third-order, and fourth-order force constants to obtain anharmonic potentials and calculate the potential energy as a function of atomic displacement. My goal is to generate a plot similar to the one in the paper "Orbitally driven giant phonon anharmonicity in SnSe" (Nature Phys. 2015).
![image](https://github.com/user-attachments/assets/2b808e50-7c45-4434-8c14-6308aeed0d18)

Could you please guide me on how to accomplish this using ALAMODE? Any advice or references to relevant documentation would be greatly appreciated.

Thank you very much for your assistance

**Comment by ttadano** [comment]
Please use https://github.com/ttadano/alamode/blob/develop/tools/calcpes.py.

For the details of the usage, please issue `python ${ALAMODE_ROOT}/tools/calcpes.y -h`

**Comment by LiuHaoXuan0403** [comment]
Dear Dr. Tadano,

Thank you for your prompt response and the guidance on using calcpes.py. I have started working with the script, and while the formats of other input files like SPOSCAR are clear, I am unsure about the disp file.

The explanation states that it is a "File containing the displacements in units of Bohr." Could you please clarify the format of this file and how I should go about creating it?

I appreciate your help.

Best regards

**Reply by ttadano** [reply]
The format is similar to 'DFSET', but only the first three columns are required.

**Reply by LiuHaoXuan0403** [reply]
Dear Dr. Tadano,

Thank you for clarifying the disp file format. However, I’ve encountered another issue. I keep receiving the error: "The number of translations in the force constant file is inconsistent."

I have tried different SPOSCAR files, including the supercell used to calculate force constants, the supercell of the primitive cell, and the supercell of conventional cell, but the error persists. Could you please advise on the correct criteria for selecting the SPOSCAR file to avoid this issue?

Your assistance is greatly appreciated.

Best regards,

**Reply by ttadano** [reply]
The `SPOSCAR` file passed to the `--structure` option must be the supercell (without displacements) used for the force constant calculations. 

This script is not well tested, so the issue you face may be due to a bug in the code.
If you need further assistance, please share the input files with me.

**Reply by Zhennan-Lin** [reply]
Dear Dr. Tadano,

I hope this email finds you well. Recently, while attempting to execute calpes.py in ALAMODE, I encountered an error that I have been unable to resolve. Below are the details of the issue:Traceback (most recent call last):
  File "calpes.py", line 117, in <module>
    main()
  File "calpes.py", line 101, in main
    taylor.set_forceconstants(fcs_dic, parser.primitive_cell)
  File "taylor.py", line 65, in set_forceconstants
    self._set_taylor_indices()
  File "taylor.py", line 223, in _set_taylor_indices
    atom_indices_tmp[i, j] = atom_indices_super[atom_indices[j]][tuple(shift_fcs[j - 1])]
KeyError: (0, 3, 0)
